pilot/pkg/serviceregistry/util/xdsfake/updater.go

```diff
@@ -105,6 +107,8 @@ func (fx *Updater) EDSUpdate(c model.ShardKey, hostname string, ns string, entry
 	default:
 	}
 	if fx.Delegate != nil {
+		fx.mu.Lock()
+		defer fx.mu.Unlock()
```

Comment:
so this makes EDSUpdate run in series in test, how about the real controller?

Reply:
That is the point I was making. The existing lock (before this change) was doing the same?

Reply:
If yes, I will update the comments and revert the lock?

Reply:
Haven't looked at the details yet, but i think the test should be consistent with pilot

Reply:
@ramaraochavali After a closer look, i think i get the root cause of the failure now. The endpointslice handler is asyn run by the queue. While the workloadInstanceHandler is run it the test synchronously, and the failure occurs when the workloadInstanceHandler run before.

From this, i am almost sure we must run `updateEDS` in serial.

Reply:
That is what I thought as well. So this `updateEDS` sequential run is needed only when `EnableK8SServiceSelectWorkloadEntries` is enabled? Should I change it to use read lock only if it is not enabled? Or should I just leave it as is and add comments so that we do not loose this information in future?

Reply:
I think it is not needed to make this trouble because updateEDS is called in serial if EnableK8SServiceSelectWorkloadEntries not enabled.as it is called by the queue.